### PR TITLE
genericapiserver: more dependency cutoffs

### PR DESCRIPTION
**Pull request by sttts**
- cut-off pkg/api.Resource and friends - lgtm
- authn plugins -> k8s.io/apiserver - 
- webhook authz plugin -> k8s.io/apiserver - lgtm
- ~~pkg/cert -> k8s.io/apimachinery (will rebase on @deads2k's PR also moving it)~~
- split pkg/config into kubelet config merger and flags - lgtm
- split feature gate between generic apiserver and kube - lgtm
- move pkg/util/flag into k8s.io/apiserver - lgtm

### CODE REVIEW

**Comment by deads2k** [comment]
`genericapiserver: cut off pkg/api.Resource dependency` lgtm

File: pkg/kubeapiserver/authenticator/config.go [hunk 1]
```diff
@@ -30,15 +30,15 @@ import (
 	"k8s.io/apiserver/pkg/authentication/request/x509"
 	"k8s.io/apiserver/pkg/authentication/token/tokenfile"
 	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/plugin/pkg/authenticator/password/keystone"
```
Comment (deads2k) [inline pkg/kubeapiserver/authenticator/config.go]: I wasn't expecting all these to move.  We have an apiserver dependency or is it just kubeapiserver?

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go [hunk 2]
```diff
@@ -24,13 +24,11 @@ import (
...
 	"k8s.io/apiserver/pkg/authentication/user"
 	"k8s.io/apiserver/pkg/util/cache"
+	"k8s.io/apiserver/pkg/util/webhook"
 	authenticationclient "k8s.io/client-go/kubernetes/typed/authentication/v1beta1"
 	authentication "k8s.io/client-go/pkg/apis/authentication/v1beta1"
 
-	"k8s.io/apiserver/pkg/util/webhook"
-
 	_ "k8s.io/client-go/pkg/apis/authentication/install"
-	_ "k8s.io/kubernetes/pkg/apis/authentication/install"
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go]: good change

File: pkg/kubeapiserver/authenticator/config.go (hunk 1 above)
Comment (sttts) [inline pkg/kubeapiserver/authenticator/config.go]: keystone is simple, no deps left

**Comment by deads2k** [comment]
`genericapiserver: move authn plugins into k8s.io/apiserver` - did you move all the authentication plugins for completeness in case someone else wanted to use them?  I could buy that, but we don't have direct dependency on them, right?

File: pkg/kubeapiserver/authenticator/config.go (hunk 1 above)
Comment (sttts) [inline pkg/kubeapiserver/authenticator/config.go]: authorizers have more dependencies. Only moved webhook there.

**Comment by sttts** [comment]
@deads2k yes, they are optional, but independent enough to not harm, if you really build your kube-independent apiserver.

**Comment by deads2k** [comment]
General comment on the authentication packages.  I think of plugins as optional and I don't think the webhook is optional and these are registering themselves for activation like normal plugins, so I'm not sure calling them plugins is appropriate.  Certainly not blocking, but of note.

**Comment by sttts** [comment]
Right about "plugin". Didn't want to change that here. Maybe for a follow-up.

**Comment by deads2k** [comment]
`pkg/util/cert: split and move generic code to k8s.io/apimachinery` - Did this split client and server stuff?  I don't think util/cert belongs in apimachinery.  Its not a great fit for client-go either, but the client already has a lot of cruft.  I'd like to try it in client-go first.  A copy is even an option later if there's little overlap.

**Comment by deads2k** [comment]
`genericapiserver: move authz webhook plugins into k8s.io/apiserver` lgtm

**Comment by deads2k** [comment]
`pkg/util: move flags from pkg/util/config to pkg/util/flags` - fine

File: pkg/api/validation/validation.go [hunk 3]
```diff
@@ -2034,7 +2035,7 @@ func ValidateAppArmorPodAnnotations(annotations map[string]string, spec *api.Pod
 		if !strings.HasPrefix(k, apparmor.ContainerAnnotationKeyPrefix) {
 			continue
 		}
-		if !utilflag.DefaultFeatureGate.AppArmor() {
+		if !utilfeature.DefaultFeatureGate.Enabled(features.AppArmor) {
```
Comment (deads2k) [inline pkg/api/validation/validation.go]: This is really ugly and generally wrong.  Not new in this pull but the link here should be snipped.  This is an admission choice, not a general validation choice.

**Comment by deads2k** [comment]
`pkg/flag: make feature gate extensible and split between generic and` - lgtm.  Probably want an issue for the bad validation.

File: pkg/api/validation/validation.go (hunk 3 above)
Comment (sttts) [inline pkg/api/validation/validation.go]: Will add a TODO. Didn't look at the meaning of the gates.

File: pkg/api/validation/validation.go [hunk 4]
```diff
@@ -2056,6 +2056,7 @@ func ValidateAppArmorPodAnnotations(annotations map[string]string, spec *api.Pod
 		if !strings.HasPrefix(k, apparmor.ContainerAnnotationKeyPrefix) {
 			continue
 		}
+		// TODO: this belongs to admission, not general pod validation:
```
Comment (sttts) [inline pkg/api/validation/validation.go]: @timstclair ^^

**Comment by sttts** [comment]
@deads2k added "move pkg/util/flag into k8s.io/apiserver". Ok with that?

**Comment by deads2k** [comment]
> @deads2k added "move pkg/util/flag into k8s.io/apiserver". Ok with that?

ick, but yes.

**Comment by sttts** [comment]
@k8s-bot bazel test this

**Comment by sttts** [comment]
@k8s-bot test this

**Comment by sttts** [comment]
Overriding approver label as this was lgtm'ed long before approvers were activated, caught in rebase hell getting through the queue.

**Comment by sttts** [comment]
@k8s-bot cvm gce e2e test this

**Comment by sttts** [comment]
@k8s-bot cri e2e test this

**Comment by sttts** [comment]
@k8s-bot cri e2e test this

**Comment by sttts** [comment]
cvm gke e2e test this

**Comment by sttts** [comment]
@k8s-bot cvm gke e2e test this

**Comment by sttts** [comment]
Rebase hell 😱

**Comment by sttts** [comment]
@k8s-bot unit test this

**Comment by sttts** [comment]
@k8s-bot unit test this

**Comment by sttts** [comment]
@k8s-bot cri node e2e test this

**Comment by sttts** [comment]
@k8s-bot bazel test this

**Comment by deads2k** [comment]
> Jenkins verification failed for commit f540fbf. Full PR test history.

@sttts keep your pull away from mine, just in case it contagious! :)

**Comment by sttts** [comment]
Looking forward to yet another rebase tomorrow morning. Trying to understand the flake, maybe this:

```
hack/make-rules/../../hack/verify-godeps.sh: line 68: KEEP_TMP: unbound variable
```

@eparis ^^

**Pull request merged**